modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/poll/PollingSourceWrapper.java

```diff
@@ -267,18 +303,43 @@ private boolean passesWatermark(DefaultPollItem pollItem) {
       boolean accept = true;
       int compare;
       if (currentWatermark == null && updatedWatermark == null) {
-        updatedWatermark = itemWatermark;
+        setUpdatedWatermark(itemWatermark);
+        pollItem.getItemId().ifPresent(id -> addToIdsOnUpdatedWatermark(id));
       } else {
-        Serializable actingWatermark = currentWatermark != null ? currentWatermark : updatedWatermark;
-        compare = compareWatermarks(actingWatermark, itemWatermark, watermarkComparator);
+        compare = currentWatermark != null ? compareWatermarks(currentWatermark, itemWatermark, watermarkComparator) : -1;
         if (compare < 0) {
-          updatedWatermark = itemWatermark;
+
+          try {
+            if (itemId != null && recentlyProcessedIds.contains(itemId)) {
+              Serializable previousItemWatermark = recentlyProcessedIds.retrieve(itemId);
+              if (compareWatermarks(itemWatermark, previousItemWatermark, watermarkComparator) <= 0) {
+                accept = false;
+              }
+            } else {
+              int updatedWatermarkCompare =
+                  updatedWatermark != null ? compareWatermarks(updatedWatermark, itemWatermark, watermarkComparator) : -1;
+              if (updatedWatermarkCompare == 0) {
+                pollItem.getItemId().ifPresent(id -> addToIdsOnUpdatedWatermark(id));
+              } else if (updatedWatermarkCompare < 0) {
+                idsOnUpdatedWatermark.clear();
+                pollItem.getItemId().ifPresent(id -> addToIdsOnUpdatedWatermark(id));
```

Comment:
do this two steps inside the `setUpdatedWatermark` to ensure that every time that the watermark is updated, items are cleared and the new id is stored.